mod/wikirate/set/type/source_import_file.rb

```diff
@@ -0,0 +1,228 @@
+  if title_card.new?
+    add_subcard title_card
+    return true
+  end
+  false
+end
+
+def handle_duplicated_source source_card, source_hash
+  # title is missing, then add the title
+  #   add company and report type to the related fields
+  #   add to warning message
+  updated = false
+  updated |= update_title_card source_card, source_hash
+  updated |= update_existing_source source_card, source_hash
+  if updated
+    slot_args = success.slot
+    msg_array = [source_hash[:row].to_s, source_card.name]
+    slot_args[:updated_sources].push(msg_array)
+  end
+  nil
+end
+
+def process_source source_hash, source_map
+  url = source_hash[:source]
+  duplicates = Self::Source.find_duplicates url
+  source_card =
+    if duplicates.any?
+      handle_duplicated_source duplicates.first.left, source_hash
+    else
+      create_source source_hash
+    end
+  source_map[url] = source_card
+end
+
+def check_duplication_within_file source_hash, source_map
+  slot_args = success.slot
+  source_url = source_hash[:source]
+  if source_map[source_url]
+    msg = [source_hash[:row].to_s, source_url]
+    slot_args[:duplicated_sources].push(msg)
+    return true
+  end
+  false
+end
+
+def valid_value_data? args
+  @import_errors = []
+  %w(wikirate_company year report_type source).each do |field|
+    add_import_error "#{field} missing", args[:row] if args[field.to_sym].blank?
+  end
+  @import_errors.empty?
+end
+
+# @return updated or created metric value card object
+def parse_source_card source, source_map
+  # check if duplicate in file
+  # if yes, add warning message and next
+  # check if source exists in db
+  # if yes,
+  #   title is missing, then add the title
+  #   add company and report type to the related fields
+  #   add to warning message
+  #   add to subcard
+  # no,
+  #   contruct the create args and add to subcards
+  args = process_data source
+  return if check_duplication_within_file args, source_map
+  return unless valid_value_data? args
+  return unless ensure_company_exists args[:company], args
+  process_source args, source_map
+end
+
+format :html do
+  include Type::MetricValueImportFile::HtmlFormat
+  def default_import_table_args args
+    args[:table_header] = ["Select", "#",  "Company in File",
+                           "Company in Wikirate", "Match",
+                           "Correction",
+                           "Year", "Report Type", "Source", "Title"]
+    args[:table_fields] = [:checkbox, :row, :file_company,
+                           :wikirate_company, :status, :correction,
+                           :year, :report_type, :source, :title]
+  end
+
+  def duplicated_value_warning_message headline, cardnames
+    message = cardnames.map do |err_row|
+      "Row #{err_row[0]}: #{err_row[1]}"
+    end.join("</li><li>")
+    msg = <<-HTML
+      <h4><b>#{headline}</b></h4>
+      <ul><li>#{message}</li> <br />
+    HTML
+    alert("warning") { msg }
```

Comment:
a bit too wasteful of variables for my taste

```ruby
alert "warning" do
  <<-HTML
    ...
  HTML
end
```

but you can (almost) never be too wasteful of methods:
```ruby
def duplicated_value_alert headline, cardname
  alert "warning" do
      <<-HTML
        <h4>#{duplicate_value_warning_message}</h4>
     HTML
  end
end
```